### PR TITLE
Replace LoadError with NameError while performing autoload failed.

### PR DESCRIPTION
**Pull request by piecehealth**
### Summary

There is some code in a third-party gems which looks like below:

```ruby
  "GemNameSpace::#{constant_string}".constantize rescue nil
```

Let's say one `constant_string` value is `User`, and we didn't have either
class `GemNameSpace::User` or class `User` yet. Then the above code worked
fine (return nil, noting raised).

After we added the class `User`, the bug occurred: the above code would raise
`LoadError` and broke our application. Since `rescue nil` only captures
`StandardError`, so before we added the `User` class, it would capture
`NameError` as a child class of `StandardError` and `return nil`. After we
added the `User` class, it raised `LoadError` and couldn't be captured by
`rescue nil` since the `LoadError` is not a child class of `StandardError`.

So if the `GemNameSpace::User` class didn't exist, `"GemNameSpace::User".constantize`
should raise `NameError` regardless which class has already exsited in our
Rails application.

### CODE REVIEW

File: actionpack/lib/abstract_controller/helpers.rb [hunk 1]
```diff
@@ -161,7 +161,7 @@ def modules_for_helpers(args)
             mod_name = file_name.camelize
             begin
               mod_name.constantize
-            rescue LoadError
+            rescue NameError
```
Comment (piecehealth) [inline actionpack/lib/abstract_controller/helpers.rb]: It seems like changed original behaviors, let me figure out a better way to fix this.

File: activesupport/test/dependencies_test.rb [hunk 2]
```diff
@@ -164,7 +164,7 @@ def test_circular_autoloading_detection
 
   def test_ensures_the_expected_constant_is_defined
     with_autoloading_fixtures do
-      e = assert_raise(LoadError) { Typo }
+      e = assert_raise(NameError) { Typo }
```
Comment (piecehealth) [inline activesupport/test/dependencies_test.rb]: maybe I should add a new test case to cover my issue.

**Comment by piecehealth** [comment]
Hi @fxn ,
Below test may describe my issue better:
```ruby
class DependenciesTest < ActiveSupport::TestCase
  # ...
  def test_should_raise_name_error_if_the_constant_doesnt_exist
    with_loading("autoloading_fixtures", "autoloading_fixtures/html") do
      assert_nothing_raised { HTML::SomeClass }
      assert_nothing_raised { ClassFolder }
      assert_raise(NameError) { ClassFolder::SomeClass.nil? }
    end
  end
 # ...
end
```

**Comment by fxn** [comment]
Let me understand the issue better.

When you say that you added the `User` class, you mean you created a file `app/models/user.rb` that defines a top-level constant `User` and reloaded? Would like to have clear reproduction steps.

Which version of Rails? Which autoloading mode if in Rails 6?

**Comment by fxn** [comment]
In a dummy Rails 6 app I have a file `app/models/user.rb` defining `User` and a directory `app/models/gem_name_space`.

This test yields `NameError`s running Rails 6 in `zeitwerk` mode:

```
% bin/rails c
Loading development environment (Rails 6.0.2.1)
irb(main):001:0> Rails.autoloaders.zeitwerk_enabled?
=> true
irb(main):002:0> "GemNameSpace::User".constantize
Traceback (most recent call last):
        1: from (irb):2
NameError (uninitialized constant GemNameSpace::User)
irb(main):003:0> User
=> User
irb(main):004:0> "GemNameSpace::User".constantize
Traceback (most recent call last):
        1: from (irb):4
NameError (uninitialized constant GemNameSpace::User)
```

and same result in `classic` mode:

```
% bin/rails c
Loading development environment (Rails 6.0.2.1)
irb(main):001:0> Rails.autoloaders.zeitwerk_enabled?
=> false
irb(main):002:0> "GemNameSpace::User".constantize
Traceback (most recent call last):
        1: from (irb):2
NameError (uninitialized constant GemNameSpace::User)
irb(main):003:0> User
=> User
irb(main):004:0> "GemNameSpace::User".constantize
Traceback (most recent call last):
        1: from (irb):4
```

Now, in `classic` mode, one way to get a `LoadError` is to define the expected file and DO NOT define the expected constant in it. Let's introduce a typo:

```
% cat app/models/user.rb 
class UserTypo
end
```

Now, `classic` mode raises `LoadError`:

```
% bin/rails r User  
...
Unable to autoload constant User, expected .../app/models/user.rb to define it (LoadError)
```

I find that exception dubious ([has been there for more than 10 years](https://github.com/rails/rails/commit/e86cced311539932420f9cda49d736606d106c28)), but do not know if that is what is causing what you're experiencing. Will be able to tell with more detailed reproduction steps.

Let me underline that the case above raises (a subclass of) `NameError` in Rails 6 in `zeitwerk` mode.

**Comment by piecehealth** [comment]
Hi @fxn,
You could refer to [my repo](https://github.com/piecehealth/pr38173) and the [commit](https://github.com/piecehealth/pr38173/commit/0460aeaef578478a77b916f408f73295a38f781f).

Run `bin/rails runner "User::Status".constantize` will reproduce my issue. The backtraces look like
```
.../.rvm/gems/ruby-2.6.5/gems/activesupport-5.2.4/lib/active_support/dependencies.rb:511:in `load_missing_constant': Unable to autoload constant Status, expected .../app/models/order/status.rb to define it (LoadError)

Prerequisite:
1. There is constant `Order` and a ruby file `app/models/order.rb`.
2. There is constant `User` and a ruby file `app/models/user.rb`.
3. There is constant `Order::Status` and a ruby file `app/models/order/status.rb`.
4. Neither `User::Status` nor `app/models/user/status.rb` is existed.
5. Folder `app/models/order` is in the `config.autoload_paths`. (I know it's against Rails's convention, but it shouldn't cause this issue neither.)

Both Rails 6.0.2.1 and 5.2.4.1 would raise `LoadError`. And if I remove `app/models/order/status.rb`, ` "User::Status".constantize` would raise `NameError` as expected.

I know I have corrupted  `autoload_paths` and broken Rails convention, but the behavior and error message is confused.

**Comment by fxn** [comment]
Hey, I am out now, but would like to understand a detail. You get a LoadError in Rails 6 and 5, but which autoloading mode in Rails 6?

**Comment by piecehealth** [comment]
@fxn , I tried it again in Rails 6, if I enabled "zeitwerk loader", it would raise `Zeitwerk::NameError`, the error message was: "Zeitwerk::NameError (expected file .../app/models/order/status.rb to define constant Status, but didn't)"
If I used the default loader, it would raise `LoadError`, the error message was: "LoadError (Unable to autoload constant Status, expected .../app/models/order/status.rb to define it)"

**Comment by fxn** [comment]
Squares, it is not an issue in `zeitwerk` mode, and I suspect you are hitting the case for LoadError in `classic` mode I explained before, only with the twist that your project has nested autoload paths. The problem with nested paths in `classic` mode is that a given file may define two constants, which is not deterministic.

Having nested autoload paths is a bad practice, and it is deterministic in `zeitwerk` mode (only the deepest one counts). Also note that the default autoloading mode in Rails 6 is `zeitwerk`, but you need to load the framework defaults for 6.0 in `config/application.rb`.

When I get back home I’ll look into the details and resolve what to do.

**Comment by fxn** [comment]
Yes, I confirm that is the problem.

The nested load paths add some noise to the root cause, but let me show your issue simplified. You need two files:

```
app/models/user.rb
app/models/status.rb
```

where the first one defines `User`, and the second one defines something that is not exactly `Status`, for example `StatusWithTypo` (in your example app that is `Order::Status`).

Now, `constantize` is sort of a recursive `const_get`, and therefore`"User::Status".constantize`tries to find `Status` in some `status.rb` in the autoload paths because `Object` is an ancestor of `User`. It finds one, and it turns out the expected constant is not defined in that file. As I explained previously, in `classic` mode that raises a `LoadError`.

That exception is debatable, I think a `NameError` would be better. However, this behaviour in `classic` mode was already present as early as in [Rails 1.2](https://github.com/rails/rails/blob/1-2-stable/activesupport/lib/active_support/dependencies.rb#L249). Maintenance of `classic` mode is coming to a halt, it is nowadays available in Rails 6 to ease the transition if you upgrade an existing app, but its usage is discouraged and eventually this mode will be removed. So, all in all, I don't think it is a good idea to introduce a backwards incompatible change.

My recommendation is:

1. Use Rails 6 with framework defaults (or enable `zeitwerk` mode if that is not possible.
2. If for some reason you need to stay in `classic` mode, please rescue both exceptions (which is what `constantize` can raise, actually, or, better, enable `zeitwerk` mode. If the `rescue` belongs to 3rd party code, that code is not catching all the exceptions that can be raised by that method call.
3. In any case, please do not use nested autoload paths if possible. They would have been forbidden in Zeitwerk if possible, they are supported only because they were a bad practice but technically possible in Rails.

**Pull request closed without merging**